### PR TITLE
Optimize glob evaluation

### LINKED ISSUE #6069: Glob pattern matching improvements for netcore/CPS (~1s during solution load of 400 projects)
Child of #6034.

Globs were not a hot path before with netframework, but with CPS using globs, the pattern matching can be optimized, say for example by extension first before groveling through disk. 

Below is the default set of blobs from CPS/netcore that comes with every project:

glob | excludedPatterns
-- | --
`**/*.cs` | `**/*.vssscc`, `obj\/**`, `**/*.sln`, `**/*.*proj`, `**/*.user`, `**/.*/**`, `bin\/**`, `bin\Debug\/**`, `obj\Debug\/**`
`**/*.resx` | `**/*.vssscc`, `obj\/**`, `**/*.sln`, `**/*.*proj`, `**/*.user`, `**/.*/**`, `bin\/**`, `bin\Debug\/**`, `obj\Debug\/**`
`**/*` | `**/*.vssscc`, `obj\/**`, `**/*.sln`, `**/*.*proj`, `**/*.user`, `**/.*/**`, `**/*.resx`, `bin\/**`, `bin\Debug\/**`, `**/*.cs`, `obj\Debug\/**`

### PR DESCRIPTION
**Pull request by ladipro**
Fixes #6069 

### Context
Glob expansion has been identified as one of the performance bottlenecks when building .NET Core projects. The cost is not negligible with a simple Hello World project and gets progressively worse when the project directory (transitively) contains more files, be it source files or completely unrelated files. Since glob expansion is part of project evaluation, it directly impacts Visual Studio performance as well, especially solution load performance.

### Changes Made
This PR is a collection of several changes to the globbing logic aiming to make .NET Core project evaluation faster.

- Refactor, simplify, and micro-optimize hot code paths,
- Eliminate exclude patterns early if it's clear that they don't intersect with the include,
- Optimize most common glob patterns to match without regex (building a typical .NET Core app now does not instantiate a single regex for glob matching),
- Cache the results of all filesystem enumeration calls, not just directories.

### Testing
Existing unit tests for correctness, ETL traces for performance.

Performance:

The wins in terms of the Microsoft-Build/ExpandGlob event duration range from 16% (Hello World project built with `MSBuild.exe`) to 48% (Hello World project with ~4000 files in ~600 folders, all non-.cs and non-.resx, built with `dotnet build`). These numbers were measured with incremental build on the command line, the project was a .NET Core project in all cases, just with a different build engine. Here's a table with all numbers:

Project | MSBuild.exe | dotnet build
--- | --- | ---
Hello World (no extra files) | 16% | 29%
Hello World (4000 extra files) | 35% | 48%

In terms of the overall build time, we're looking at a 1-2% improvement for a Hello World project and 8-10% for a project with thousands of extra files in the source tree.

### Notes
This PR is recommended to be reviewed commit by commit. `FileMatcher` is a complex class and there is non-trivial breaking potential as projects may depend on all kinds of corner case behavior. Which is the primary reason for not rewriting the entire thing. The most impactful change in this PR - caching files to avoid repeated filesystem API calls - is behind a change wave condition, just in case.

### CODE REVIEW

File: src/Shared/FileMatcher.cs
Comment (ladipro) [inline src/Shared/FileMatcher.cs:1050]: @cdmihai may I please get your opinion on this? Glob matching on Linux is currently sometimes case-sensitive and sometimes case-insensitive depending on which code path we happen to take in this function.

In this PR I am optimizing globbing and without the hack/workaround a couple of test cases fail on Linux because they expect certain behavior w.r.t. case sensitivity. These assumptions are encoded very explicitly:

https://github.com/dotnet/msbuild/blob/6bad8ea6a27fc23621ee8a41fd941dff01976de1/src/Shared/UnitTests/FileMatcher_Tests.cs#L112-L120

and for example:

https://github.com/dotnet/msbuild/blob/6bad8ea6a27fc23621ee8a41fd941dff01976de1/src/Shared/UnitTests/FileMatcher_Tests.cs#L219

How would you feel about fixing Linux to be consistent, so most likely case-insensitive like other platforms, by changing file enumeration calls to pass [EnumerationOptions ](https://docs.microsoft.com/en-us/dotnet/api/system.io.enumerationoptions?view=net-5.0) with `MatchCasing=false`? Technically a breaking change, yes, but the current state makes it hard to innovate and has us write ugly hacks like this one.

File: src/Shared/FileMatcher.cs
Comment (BenVillalobos) [inline src/Shared/FileMatcher.cs:1695]: Could you explain how you came to this micro-optimization?

File: src/Shared/FileMatcher.cs
Comment (BenVillalobos) [reply src/Shared/FileMatcher.cs:1050]: Not completely relevant, but a bit of a discussion happened on linux case sensitivity here: https://github.com/dotnet/msbuild/issues/5658

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:1716]: The original code had to do 4 comparisons here: 2 to make sure that `inputChar` is in range and 2 to make sure that `patternChar` is in the corresponding range also. Then it compared them knowing that pattern, being lower-case, needs to have its ASCII value 32 more than input for them to match.

I simplified the condition to check only one side, in this case `inputChar`. And instead of adding/subtracting 32 I am setting/clearing the corresponding bit, making the check work regardless of what the value of `patternChar` is.

Example: If `inputChar` is `A` and `patternChar` is `A` then clearing the bit does nothing as upper case characters don't have it set so the result will be `true`, i.e. they are the same. If `inputChar` is `A` and `patternChar` is `a` then clearing the bit on `patternChar` turns it into `A` and again, the check works.

This line could equivalently be written as:

```C#
return (inputChar | 0x20) == patternChar;
```

A micro benchmark for `CompareIgnoreCase` showed ~10% perf improvement with this change in the common case where the input are letters from the English alphabet.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:2123]: Yes, I think that there are opportunities to optimize by using `Span<>` in this code, although most of it is not low-hanging fruit and doesn't have to addressed with priority. This particular line, for example, runs only a handful of times per project so no biggie if it allocates.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:1671]: Yes, all calls were passing true. After implementing a few more optimizations, I suspect I may need to put the parameter back, though, for back compat on Linux.

File: src/Build/Evaluation/Context/EvaluationContext.cs
Comment (donJoseLuis) [inline src/Build/Evaluation/Context/EvaluationContext.cs:51]: very minor: could this be readonly as it's instantiated in the .ctor?
However, readonly gives the impression of immutability after instantiation, not the case for collections.

File: src/Build/Globbing/MSBuildGlob.cs
Comment (donJoseLuis) [inline src/Build/Globbing/MSBuildGlob.cs:125]: not a change in the PR.  It's difficult to infer normalizedInput's type.

File: src/Build/Globbing/MSBuildGlob.cs
Comment (donJoseLuis) [inline src/Build/Globbing/MSBuildGlob.cs:130]: minor: is 4 out parameters better than a new encapsulation of the same state?

File: src/Build/Globbing/MSBuildGlob.cs
Comment (donJoseLuis) [inline src/Build/Globbing/MSBuildGlob.cs:145]: this would allow white space to pass the test, right?

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (donJoseLuis) [inline src/Build/Utilities/FileSpecMatchTester.cs:24]: minor: no need to assert other parameters validity?

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (donJoseLuis) [inline src/Build/Utilities/FileSpecMatchTester.cs:62]: minor: counterintuitive type inference.

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (donJoseLuis) [inline src/Build/Utilities/FileSpecMatchTester.cs:63]: minor: consider referencing the System.StringComparison namespace

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:1079]: Here's a simple example where we are inconsistent today.

Source tree:
```
.
├── abc
│   ├── File
│   └── file
└── repro.proj
```

repro.proj:
```xml
<Project>
  <ItemGroup>
    <Item1 Include="a*/f*" />
    <Item2 Include="abc/f*" />
  </ItemGroup>

  <Target Name="Build">
    <Message Text="@(Item1)" />
    <Message Text="@(Item2)" />
  </Target>
</Project>
```

```sh
$ dotnet build repro.proj /v:n
```

Prints:
```
  abc/File;abc/file
  abc/file
```

i.e. the first glob matched both ```file``` and ```File``` while the second one did not.

File: src/Build/Evaluation/Context/EvaluationContext.cs
Comment (ladipro) [reply src/Build/Evaluation/Context/EvaluationContext.cs:51]: It's a property and defining only `get;` actually makes it read-only. The `readonly` keyword would not be valid here.

File: src/Build/Globbing/MSBuildGlob.cs
Comment (ladipro) [reply src/Build/Globbing/MSBuildGlob.cs:145]: No, `IsNullOrEmpty` really tests only for `null` and empty string. There's also `IsNullOrWhiteSpace` which is more permissive.

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:64]: Done, thanks!

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:63]: Agreed and fixed, thanks!

File: src/Build/Globbing/MSBuildGlob.cs
Comment (ladipro) [reply src/Build/Globbing/MSBuildGlob.cs]: Fixed.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:1672]: It looks like I ended up not needing the parameter.

File: src/Build/Evaluation/Context/EvaluationContext.cs
Comment (ladipro) [inline src/Build/Evaluation/Context/EvaluationContext.cs:6]: I am switching over from `ImmutableArray<string>` to `IReadOnlyList<string>` because the former is ~4x more expensive to construct from an `IEnumerable`. We don't really use the immutable functionality here as we never "mutate" the collection after it's created.

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [reply src/Shared/FileMatcher.cs:1080]: Case sensitivity is part of a long standing undesigned area of MSBuild: path representation. See #781 and #1024. It's pretty much a can of worms.

However way we change it, we'll probably break *somebody* and [as noted before](https://github.com/dotnet/msbuild/issues/1024#issuecomment-549564420), it's probably of minimal user benefit.
That being said, it would be interesting to test these assumptions and hardcode casing to case insensitive, provided that an escape hatch to old behaviour is provided. Unfortunately the escape hatch would keep the messy code, but we could at least give it a year or so to hear any complaints before deleting it.

If you do choose to embark on this journey, I'd prefer the following:
- design it and take the other aspects of #781 and #1024 into account, even if just as placeholders for future design. The make it an official spec on docs.microsoft.com on how msbuild handles file system configurations.
- make all the msbuild code base uniform to the new spec, not just globbing
- have a central place in the codebase that decides file system configuration policies (similar to how [BuildEnvironment](https://github.com/dotnet/msbuild/blob/f0eebf2872d76ab0cd43fdc4153ba636232b222f/src/Shared/BuildEnvironmentHelper.cs#L497) centralizes msbuild file and assembly structure) instead of having it duplicated all over the place. One policy would be the `Legacy` policy, which is the behaviour today. Hopefully we can delete it in 1-2 years if nothing really critical breaks.

Comment (ladipro) [reply src/Shared/FileMatcher.cs:1080]: Thank you for the details. It sounds like changing the behavior should be done separately so I am keeping the Linux workaround and this PR overall strives for 100% back compat.

File: src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs
Comment (ladipro) [inline src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs:341]: This is an unintentional change, will remove next time I update the PR.

**Review by rokonec — APPROVED** [review]

File: src/Shared/FileMatcher.cs
Comment (Forgind) [inline src/Shared/FileMatcher.cs:2135]: I'm also wondering how common the case you're looking at here is. My gut tells me it's less common than the last case you considered. Also, could we allow slashes in the interior if we also fix them to be a common form and disallow wild cards?

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (Forgind) [inline src/Build/Utilities/FileSpecMatchTester.cs:58]: This feels a little like a hammer to me. Maybe a separate method that handles small cases without slashes or complicated wildcards might do? If not, this should work.

File: src/Shared/FileMatcher.cs
Comment (Forgind) [inline src/Shared/FileMatcher.cs:1070]: tiny nit:
If you give this true as a default value, you don't have to reassign it for the first case.

File: src/Shared/FileMatcher.cs
Comment (Forgind) [inline src/Shared/FileMatcher.cs:125]: Nit:
Use $" to prevent intermediate allocation?

File: src/Shared/FileMatcher.cs
Comment (Forgind) [inline src/Shared/FileMatcher.cs:1534]: RegularExpressionFromFileSpec starts by checking whether IsLegalFileSpec(...). It's private and hardly used. Could we pull that out here so we don't check that twice?

Comment (Forgind) [reply src/Shared/FileMatcher.cs:1534]: Actually, I think you can just delete it. It seems to be checked before every call.

**Review by Forgind — COMMENTED** [review]
This looks good to me overall, thanks!

The only part of this that seemed like a big (potential) gain was a heuristic. Do you know how helpful this was for users?

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [inline src/Shared/FileMatcher.cs:2459]: It would be nice to add / find existing tests that hits this condition on a couple of cases:
- non empty text after last wildcard
- empty text after last wildcard

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [inline src/Shared/FileMatcher.cs:799]: What if there's multiple directory patterns like `**/{pattern1}/**/{pattern2}/**`

Comment (cdmihai) [reply src/Shared/FileMatcher.cs:799]: Oh, as directory traversal descends one FilesSearchData instance will have it set to `pattern1`, then a later instance will have it set to `pattern2`, right? Remembering this code brings up some PTSD :P

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [inline src/Shared/FileMatcher.cs:943]: Add / find tests that hit the newly added conditions.

**Review by cdmihai — APPROVED** [review]
LGTM pending ensuring tests cover the new conditions / code paths

File: src/Build/Globbing/MSBuildGlob.cs
Comment (ladipro) [reply src/Build/Globbing/MSBuildGlob.cs:128]: Unfortunately cannot do what you're suggesting as `GetRegexMatchInfo` lives in src/shared, `MSBuildGlob` in src/Build, and `MatchInfoResult` is a public type so it can't be easily moved. I'd say 4 output parameters, while a minor readability hit, is borderline acceptable. There are two calls to `GetRegexMatchInfo`, both populating their own structure with the returned data. Adding another representation just for the purpose of returning it from the method would IMO be an overkill.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs]: Thank you, that's a great point! I have pushed your suggestion, just with `inputCharLower = ` as it's a bit easier to read than `x & ~y`.

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:26]: Yes, added a `currentDirectory` validity check as well. Thank you.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:125]: The compiler generates a call to `String.Concat(string, string, string)` in both cases: https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMCADJgLIIAUF1AzggDTZWbNwCUhxUkJgB2bgkwBqTACIA3DKnc4cwSQC+qNWU7UacRl1Ycmy/kRRDhYgCQyCrdXIdx187ZpTqgA

In cases where most of the expression is literal, I find `$"` useful, e.g. `$"Hello {name}, how are you?"`. Otherwise, when there's only a little literal string and there's no perf impact, I prefer the good old `+`.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:1534]: The check is already gone, were you maybe looking at an older version of the PR. Thanks!

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs]: Actually, let's remove the local altogether and use the existing `NeedsToProcessEachFile` field on the struct.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs]: Thank you, yes, that's a great point. I have re-added materialization calls except for places where keeping the `IEnumerable<>` clearly has benefits.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:2445]: The work here was driven by what the .NET SDK uses when building common applications. The full list is in the description of #6069. I'm not sure what you mean by `*.`, is it literally `*.`, so files ending with a dot?

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:2125]: In .NET SDK, this is the `**/.*/**` pattern used for excluding files from hidden directories, such as `.git`. If we allow slashes in the interior, it's no longer going to be "do we have a directory on the path matching {some pattern}?" and things get more complicated because we can't simply check a directory when we're descending into it. I'm actually fine letting this fall into the slow regex case because it's not used in the SDK.

File: src/Build/Utilities/FileSpecMatchTester.cs
Comment (ladipro) [reply src/Build/Utilities/FileSpecMatchTester.cs:59]: `_filenamePattern` may have complicated wildcards so using `FileMatcher.IsMatch` is a good match (pun intended), as far as I can tell. Feel free to unresolve this comment if I misunderstood.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:798]: I am not optimizing `**/{pattern1}/**/{pattern2}/**` in this PR. This one I'm fine being handled by the slow regex path because it's not used by the .NET SDK and overall doesn't feel like something common (I have no data to back it up, just a gut feeling).

I think that you're right that it could be done without regex'es, although I'm sure there are many gotchas that prevent it from working simply by peeling the path components one by one as we're descending. For example, the `**` should not do eager matching so you still generally need a "lookahead". It may happen that you've descended multiple levels and everything matched but then you find a mismatch that could have been prevented by going back and changing your mind about a component matching against `**` or not.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:2449]: Great point, added two new `GetFilesComplexGlobbingMatchingInfo`'s exercising this code path.

File: src/Shared/FileMatcher.cs
Comment (Forgind) [reply src/Shared/FileMatcher.cs:2445]: Sorry—it formatted the markdown to look like *., but that wasn't what I'd intended to put.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:942]: This is well covered with `RecursiveDirRecursive` and more.

https://github.com/dotnet/msbuild/blob/f0eebf2872d76ab0cd43fdc4153ba636232b222f/src/Shared/UnitTests/FileMatcher_Tests.cs#L1021-L1033

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [reply src/Shared/FileMatcher.cs:798]: Might help to make the comment more explicit that the field is only applicable for that particular optimization and it's not part of the general flow

File: src/Shared/FileMatcher.cs
Comment (cdmihai) [inline src/Shared/FileMatcher.cs:2462]: no Hoho :)

**Review by cdmihai — APPROVED** [review]

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs:798]: Done, thank you.

File: src/Shared/FileMatcher.cs
Comment (ladipro) [reply src/Shared/FileMatcher.cs]: And now you know how I debug my code :D Removed!

**Review by ladipro — COMMENTED** [review]
Thank you all for the feedback. I will squash PR review changes into the right commits before merging to make git history look reasonable.

**Comment by ForNeVeR** [comment]
Hey, folks, could this be related to https://github.com/dotnet/sdk/issues/17976? Could someone please take a look?

**Pull request merged**
